chain-signatures/contract/src/lib.rs

```diff
@@ -94,29 +115,44 @@ impl SignatureRequest {
 #[derive(BorshDeserialize, BorshSerialize, Debug)]
 pub struct MpcContract {
     protocol_state: ProtocolContractState,
-    pending_requests: LookupMap<SignatureRequest, Option<SignatureResponse>>,
+    pending_requests: LookupMap<SignatureRequest, Option<CryptoHash>>,
     request_counter: u32,
+    yield_resume_requests: LookupMap<u64, YieldResumeRequest>,
+    next_available_yield_resume_request_index: u64,
 }
 
 impl MpcContract {
-    fn add_request(&mut self, request: &SignatureRequest, result: &Option<SignatureResponse>) {
+    fn add_request(
+        &mut self,
+        request: &SignatureRequest,
+        yield_resume_data_id: &Option<CryptoHash>,
+    ) {
         if self.request_counter > 8 {
             env::panic_str("Too many pending requests. Please, try again later.");
         }
         if !self.pending_requests.contains_key(request) {
             self.request_counter += 1;
         }
-        self.pending_requests.insert(request, result);
+        self.pending_requests.insert(request, yield_resume_data_id);
     }
 
-    fn remove_request(&mut self, payload: &SignatureRequest) {
-        self.pending_requests.remove(payload);
-        self.request_counter -= 1;
+    fn add_yield_resume_request(&mut self, index: u64, yield_resume_request: YieldResumeRequest) {
+        self.yield_resume_requests
+            .insert(&index, &yield_resume_request);
     }
 
-    fn add_sign_result(&mut self, payload: &SignatureRequest, signature: SignatureResponse) {
-        if self.pending_requests.contains_key(payload) {
-            self.pending_requests.insert(payload, &Some(signature));
+    fn remove_request_by_yield_resume_index(&mut self, index: u64) {
+        if let Some(YieldResumeRequest {
+            data_id: _,
+            account_id: _,
+            signature_request,
+        }) = self.yield_resume_requests.get(&index)
+        {
+            self.yield_resume_requests.remove(&index);
+            self.pending_requests.remove(&signature_request);
```

Comment:
We can directly just call `remove` instead of calling `get`:
```suggestion
        if let Some(YieldResumeRequest {
            data_id: _,
            account_id: _,
            signature_request,
        }) = self.yield_resume_requests.remove(&index)
        {
            self.pending_requests.remove(&signature_request);
```